shell/platform/windows/windows_lifecycle_manager.cc

```diff
@@ -0,0 +1,134 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "windows_lifecycle_manager.h"
+
+#include <TlHelp32.h>
+#include <WinUser.h>
+#include <Windows.h>
+#include <tchar.h>
+
+#include "flutter/shell/platform/windows/flutter_windows_engine.h"
+
+namespace flutter {
+
+WindowsLifecycleManager::WindowsLifecycleManager(FlutterWindowsEngine* engine)
+    : engine_(engine) {}
+
+WindowsLifecycleManager::~WindowsLifecycleManager() {}
+
+void WindowsLifecycleManager::Quit(UINT exit_code, HWND hwnd) {
+  if (hwnd == nullptr) {
+    ::PostQuitMessage(exit_code);
+  } else {
+    sent_close_messages_[hwnd]++;
+    PostMessage(hwnd, WM_CLOSE, 0, 0);
+  }
+}
+
+bool WindowsLifecycleManager::WindowProc(HWND hwnd,
+                                         UINT msg,
+                                         WPARAM wpar,
+                                         LPARAM lpar,
+                                         LRESULT* result) {
+  switch (msg) {
+    case WM_CLOSE:
+      auto itr = sent_close_messages_.find(hwnd);
+      if (itr != sent_close_messages_.end()) {
+        if (itr->second == 1) {
+          sent_close_messages_.erase(itr);
+        } else {
+          sent_close_messages_[hwnd]--;
+        }
+        return false;
+      }
+      if (IsLastWindowOfProcess()) {
+        engine_->RequestApplicationQuit(AppExitType::cancelable, 0, hwnd);
+      }
+      return true;
```

Comment:
Is this right? Should we leave the `WM_CLOSE` message unhandled if this isn't the last window of the process?